qiskit/transpiler/passes/optimization/inverse_cancellation.py

```diff
@@ -0,0 +1,113 @@
+
+
+class Cancellation(TransformationPass):
+    """Cancel back-to-back `gates` in dag."""
+
+    def __init__(self, gates_to_cancel: List[Union[Gate, Tuple[Gate, Gate]]]):
+        """Initialize gates_to_cancel.
+
+        Args:
+            gates_to_cancel: list of gates to cancel 
+        """
+        for gates in gates_to_cancel:
+            if isinstance(gates, Gate):
+                
+                if gates != gates.inverse():
+                    raise TranspilerError('Gates are not self inverse')
+            else: 
+                if isinstance(gates, tuple):
+                    if len(gates) != 2:
+                        raise TranspilerError('Too many or too few inputs')
+                    else:
+                        if gates[0] != gates[1].inverse():
+                            raise TranspilerError('Gates are not inverse')
+                else:
+                    raise TranspilerError('Argument does not take input type')
+
+        self.gates_to_cancel = gates_to_cancel
+        super().__init__()
+
+    def run(self, dag: DAGCircuit):
+        """Run the Cancellation pass on `dag`.
+
+        Args:
+            dag: the directed acyclic graph to run on.
+
+        Returns:
+            DAGCircuit: Transformed DAG.
+        """
+        self_inverse_gates = []
+        inverse_gate_pairs = []
+
+        for gates in self.gates_to_cancel:
+            if isinstance(gates, Gate):
+                self_inverse_gates.append(gates)
+            else:
+                inverse_gate_pairs.append(gates)
+
+        dag = self._run_on_self_inverse(dag, self_inverse_gates)
+        return self._run_on_inverse_pairs(dag, inverse_gate_pairs)
+
+    def _run_on_self_inverse(self, dag: DAGCircuit, self_inverse_gates: List[Gate]):
+        """
+        Run self-inverse gates on `dag`.
+
+        Args:
+            dag: the directed acyclic graph to run on.
+            self_inverse_gates: list of gates who cancel themeselves in pairs
+
+        Returns:
+            DAGCircuit: Transformed DAG.
+        """
+        gate_cancel_runs = dag.collect_runs([gate.name for gate in self_inverse_gates])
+        for gate_cancel_run in gate_cancel_runs:
+            if len(gate_cancel_run) % 2 == 0:
+                dag.remove_op_node(gate_cancel_run[0])
+            for node in gate_cancel_run[1:]:
+                dag.remove_op_node(node)
+        return dag
+
+    def _run_on_inverse_pairs(self, dag: DAGCircuit, inverse_gate_pairs: List[Tuple[Gate, Gate]]):
+        """
+        Run inverse gate pairs on `dag`.
+        
+        Args:
+            dag: the directed acyclic graph to run on.
+            inverse_gate_pairs: list of gates with inverse angles that cancel each other.
+
+        Returns:
+            DAGCircuit: Transformed DAG.
+        """
+        for pair in inverse_gate_pairs:
+            gate_cancel_runs = dag.collect_runs([pair[0].name]) 
+            for dag_node in gate_cancel_runs:
```

Comment:
Whoops, I think I suggested this name, but since `dag_node` is multiple nodes, let's rename this `dag_nodes` plural. Reminder you can highlight it and hold cmd+d to multi-select in VSCode